nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/UpdateDatabaseTable.java

```diff
@@ -0,0 +1,606 @@
+                        // but it's possible it might not have a message. This handles that by logging
+                        // the name of the class thrown.
+                        Throwable c = e.getCause();
+                        if (c != null) {
+                            session.putAttribute(flowFile, "record.error.message", (c.getLocalizedMessage() != null) ? c.getLocalizedMessage() : c.getClass().getCanonicalName() + " Thrown");
+                        } else {
+                            session.putAttribute(flowFile, "record.error.message", e.getClass().getCanonicalName() + " Thrown");
+                        }
+                        session.transfer(flowFile, REL_FAILURE);
+                        return;
+                    }
+
+                }
+                attributes.put(ATTR_OUTPUT_TABLE, tableName);
+                flowFile = session.putAllAttributes(flowFile, attributes);
+                session.getProvenanceReporter().invokeRemoteProcess(flowFile, getJdbcUrl(connection));
+                session.transfer(flowFile, REL_SUCCESS);
+            }
+        } catch (IOException | SQLException e) {
+            flowFile = session.putAttribute(flowFile, ATTR_OUTPUT_TABLE, tableName);
+            log.error("Exception while processing {} - routing to failure", new Object[]{flowFile}, e);
+            session.transfer(flowFile, REL_FAILURE);
+        } catch (DiscontinuedException e) {
+            // The input FlowFile processing is discontinued. Keep it in the input queue.
+            getLogger().warn("Discontinued processing for {} due to {}", new Object[]{flowFile, e}, e);
+            session.transfer(flowFile, Relationship.SELF);
+        } catch (Throwable t) {
+            throw (t instanceof ProcessException) ? (ProcessException) t : new ProcessException(t);
+        }
+    }
+
+    private synchronized OutputMetadataHolder checkAndUpdateTableSchema(final Connection conn, final DatabaseAdapter databaseAdapter, final RecordSchema schema,
+                                                                        final String catalogName, final String schemaName, final String tableName,
+                                                                        final boolean createIfNotExists, final boolean translateFieldNames, final boolean updateFieldNames,
+                                                                        final boolean quoteTableName, final boolean quoteColumnNames) throws IOException {
+        // Read in the current table metadata, compare it to the reader's schema, and
+        // add any columns from the schema that are missing in the table
+        try (Statement s = conn.createStatement()) {
+            // Determine whether the table exists
+            TableSchema tableSchema = null;
+            try {
+                tableSchema = TableSchema.from(conn, catalogName, schemaName, tableName, translateFieldNames, null, getLogger());
```

Comment:
I notice we're passing null here to `updateKeys` -- in using this processor in a flow, I found that I wanted a way to specify primary keys in the table that was created.  Is there another way to do this in `UpdateDatabaseTable`?  If not, what do you think about adding a property to identify primary keys in the schema during create?

Reply:
Sounds good, there's one in PutDatabaseRecord but it does something different. In this case we'd use it for specifying primary keys during table creation as you said. Great catch!